Dockerfile

```diff
@@ -1,35 +1,40 @@
 FROM ubuntu:latest as smartdns-builder
 LABEL previous-stage=smartdns-builder
-COPY . /smartdns/
+
+# prepare builder
+ARG OPENSSL_VER=1.1.1f
 RUN apt update && \
-    apt install -y perl wget make musl-tools musl-dev && \
-    OPENSSL_VER=1.1.1f && \
-    mkdir /build -p && \
+    apt install -y perl curl make musl-tools musl-dev && \
     ln -s /usr/include/linux /usr/include/$(uname -m)-linux-musl && \
     ln -s /usr/include/asm-generic /usr/include/$(uname -m)-linux-musl && \
     ln -s /usr/include/$(uname -m)-linux-gnu/asm /usr/include/$(uname -m)-linux-musl && \
-    cd /build && \
-    wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/openssl_${OPENSSL_VER}.orig.tar.gz && \
-    tar xf openssl_${OPENSSL_VER}.orig.tar.gz && \
-    cd openssl-${OPENSSL_VER} && \
+    \
+    mkdir -p /build/openssl && \
+    cd /build/openssl && \
+    curl -sSL http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/openssl_${OPENSSL_VER}.orig.tar.gz | tar --strip-components=1 -zxv && \
+    \
     export CC=musl-gcc && \
     if [ "$(uname -m)" = "aarch64" ]; then \
         ./config --prefix=/opt/build no-tests -mno-outline-atomics ; \
     else \ 
-        ./config --prefix=/opt/build no-tests; \
+        ./config --prefix=/opt/build no-tests ; \
     fi && \
     make all -j8 && make install_sw && \
-    cd /smartdns && \
+    cd / && rm -rf /build
+
+# do make
+COPY . /build/smartdns/
+RUN cd /build/smartdns && \
     export CFLAGS="-I /opt/build/include" && \
     export LDFLAGS="-L /opt/build/lib" && \
     sh ./package/build-pkg.sh --platform linux --arch `dpkg --print-architecture` --static && \
-    mkdir /release -p && \
-    cd /smartdns/package && tar xf *.tar.gz && \
-    cp /smartdns/package/smartdns/etc /release/ -a && \
-    cp /smartdns/package/smartdns/usr /release/ -a && \
-    chmod +x /release/etc/init.d/smartdns && \
-    mkdir /release/var/log/ /release/var/run/ -p && \
-    rm -fr /build /smartdns
+    \
+    ( cd package && tar -xvf *.tar.gz && chmod a+x smartdns/etc/init.d/smartdns ) && \
```

Comment:
The empty file "package/openwrt/files/etc/init.d/smartdns" should not be added

Reply:
removed

Reply:
I mean, Nothing changed in `package/openwrt/files/etc/init.d/smartdns`, do not commit this file.

Reply:
reverted with empty file removed

Reply:
Tested, there is a problem with the product. When compiling smartdns, musl-gcc is not used, and link with glibc, not musl.

I think, the useful modification is parameterized version of openssl. It is recommended to keep only the  parameterized version modification, and leave other code unchanged.

Reply:
> When compiling smartdns, musl-gcc is not used, and link with glibc, not musl.

env `CC` missing, fixed.

Reply:
Why split RUN into two RUNs?

Reply:
> make cache happy


install os-deps and compile `openssl` every time when `docker build` is unnecessary, just do `make` if src  is changed

Reply:
refer: https://www.freecodecamp.org/news/docker-cache-tutorial/